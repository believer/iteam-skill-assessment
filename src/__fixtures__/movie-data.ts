import { build, fake } from 'test-data-bot'
import { SearchMovie, MovieType, Movie } from '../types'

type PartialMovie = Pick<
  Movie,
  'Actors' | 'Poster' | 'Title' | 'Type' | 'Year' | 'imdbID'
>

export const searchMovieBuilder = build<SearchMovie>('Movie').fields({
  Poster: fake(f => f.internet.url()),
  Title: fake(f => f.lorem.words()),
  Type: MovieType.Movie,
  Year: fake(f =>
    f.date
      .past()
      .getFullYear()
      .toString()
  ),
  imdbID: fake(f => f.random.alphaNumeric()),
})

export const movieBuilder = build<PartialMovie>('Movie').fields({
  Actors: 'test, test2, test3',
  Poster: fake(f => f.internet.url()),
  Title: fake(f => f.lorem.words()),
  Type: MovieType.Movie,
  Year: fake(f =>
    f.date
      .past()
      .getFullYear()
      .toString()
  ),
  imdbID: fake(f => f.random.alphaNumeric()),
})
