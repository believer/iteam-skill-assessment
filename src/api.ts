import useFetch from 'react-fetch-hook'
import { MovieResponse, SearchResults } from './types'

const API_KEY = process.env.REACT_APP_OMDB_KEY

const createRequest = (params: string) =>
  `https://www.omdbapi.com/?apikey=${API_KEY}${params}`

export const useSearch = (query: string) => {
  const { data, isLoading, error } = useFetch<SearchResults>(
    createRequest(`&s=${query}`),
    {
      depends: [query],
    }
  )

  if (!data) {
    return { movies: [], isLoading, error }
  }

  return { movies: data.Search || [], isLoading, error }
}

export const useMovie = (id?: string) => {
  const { data, isLoading, error } = useFetch<MovieResponse>(
    createRequest(`&i=${id}`),
    {
      depends: [id],
    }
  )

  if (!data) {
    return { movie: null, isLoading, error }
  }

  if (data.Error) {
    return { movie: null, isLoading, error: data.Error }
  }

  return {
    movie: data,
    isLoading,
    error,
  }
}
