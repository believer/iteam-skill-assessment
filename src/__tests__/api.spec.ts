import { useSearch, useMovie } from '../api'
import useFetch from 'react-fetch-hook'
import { searchMovieBuilder, movieBuilder } from '../__fixtures__/movie-data'

jest.mock('react-fetch-hook')

afterEach(jest.clearAllMocks)

describe('#useSearch', () => {
  test('handles null data', () => {
    ;(useFetch as jest.Mock).mockReturnValue({
      data: null,
      isLoading: false,
      error: null,
    })

    const result = useSearch('inception')

    expect(result).toMatchSnapshot()
  })

  test('search and return data', () => {
    const movieOne = searchMovieBuilder()
    const movieTwo = searchMovieBuilder()
    ;(useFetch as jest.Mock).mockReturnValue({
      data: {
        Search: [movieOne, movieTwo],
      },
      isLoading: false,
      error: null,
    })

    const result = useSearch('inception')

    expect((useFetch as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
    expect(result).toEqual({
      error: null,
      isLoading: false,
      movies: [movieOne, movieTwo],
    })
  })
})

describe('#useMovie', () => {
  test('handles null data', () => {
    ;(useFetch as jest.Mock).mockReturnValue({
      data: null,
      isLoading: false,
      error: null,
    })

    const result = useMovie('1')

    expect(result).toMatchSnapshot()
  })

  test('search and return data', () => {
    const movie = movieBuilder()
    ;(useFetch as jest.Mock).mockReturnValueOnce({
      data: movie,
      isLoading: false,
      error: null,
    })

    const result = useMovie('1')

    expect((useFetch as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
    expect(result).toEqual({
      movie,
      isLoading: false,
      error: null,
    })
  })
})
