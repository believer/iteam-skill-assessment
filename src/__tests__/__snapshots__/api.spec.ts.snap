// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#useMovie handles null data 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "movie": null,
}
`;

exports[`#useMovie search and return data 1`] = `"https://www.omdbapi.com/?apikey=test_key&i=1"`;

exports[`#useSearch handles null data 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "movies": Array [],
}
`;

exports[`#useSearch search and return data 1`] = `"https://www.omdbapi.com/?apikey=test_key&s=inception"`;
